src/drivers/intel/cavs/hda-dma.c

```diff
@@ -310,12 +312,28 @@ static int hda_dma_host_preload(struct dma *dma, struct hda_chan_data *chan)
 static void hda_dma_post_copy(struct dma *dma, struct hda_chan_data *chan,
 			      int bytes)
 {
+	uint32_t status;
+	uint64_t delay;
 	struct dma_sg_elem next = {
 			.src = DMA_RELOAD_LLI,
 			.dest = DMA_RELOAD_LLI,
 			.size = bytes
 	};
+	/* wait until DMA finish its copy */
+	status = host_dma_reg_read(chan->dma, chan->index, DGCS);
+	delay = platform_timer_get(platform_timer) +
+			 clock_ms_to_ticks(PLATFORM_DEFAULT_CLOCK, 1) *
+			 PLATFORM_HOST_DMA_TIMEOUT / 1000;
+
+	while (status & DGCS_BNE || status & DGCS_GBUSY) {
```

Comment:
we need to add direction check here, for playback, we might need to check DGCS_BF, not DGCS_BNE.

Reply:
In fact they won't be broken because of the existence of the timeout. Although I agree that flow detection is needed.